OpenRA.Mods.Common/Traits/World/PaletteFromJascFile.cs

```diff
@@ -0,0 +1,106 @@
+#region Copyright & License Information
+/*
+ * Copyright 2007-2017 The OpenRA Developers (see AUTHORS)
+ * This file is part of OpenRA, which is free software. It is made
+ * available to you under the terms of the GNU General Public License
+ * as published by the Free Software Foundation, either version 3 of
+ * the License, or (at your option) any later version. For more
+ * information, see COPYING.
+ */
+#endregion
+
+using System.Collections.Generic;
+using System.Drawing;
+using System.IO;
+using OpenRA.Graphics;
+using OpenRA.Traits;
+
+namespace OpenRA.Mods.Common.Traits
+{
+	[Desc("Load a JASC PaintShop Pro Palette .pal file.")]
+	class PaletteFromJascFileInfo : ITraitInfo
+	{
+		[FieldLoader.Require, PaletteDefinition]
+		[Desc("Palette name used internally.")]
+		public readonly string Name = null;
+
+		[Desc("If defined, load the palette only for this tileset.")]
+		public readonly string Tileset = null;
+
+		[FieldLoader.Require]
+		[Desc("Name of the file to load.")]
+		public readonly string Filename = null;
+
+		[Desc("Map listed indices to shadow. Ignores previous color.")]
+		public readonly int[] ShadowIndex = { };
+
+		public readonly bool AllowModifiers = true;
+
+		public object Create(ActorInitializer init) { return new PaletteFromJascFile(init.World, this); }
+	}
+
+	class PaletteFromJascFile : ILoadsPalettes, IProvidesAssetBrowserPalettes
+	{
+		readonly World world;
+		readonly PaletteFromJascFileInfo info;
+
+		public PaletteFromJascFile(World world, PaletteFromJascFileInfo info)
+		{
+			this.world = world;
+			this.info = info;
+		}
+
+		public void LoadPalettes(WorldRenderer wr)
+		{
+			var colors = new uint[Palette.Size];
+			using (var s = world.Map.Open(info.Filename))
+			{
+				var lines = s.ReadAllLines().GetEnumerator();
+
+				if (!lines.MoveNext() || lines.Current != "JASC-PAL" || !lines.MoveNext() || !lines.MoveNext())
+					throw new InvalidDataException("File {0} is not a valid JASC palette.".F(info.Filename));
+
+				if (int.Parse(lines.Current) != Palette.Size)
+					throw new InvalidDataException("Only {0} color palettes are supported.".F(Palette.Size));
+
+				byte r, g, b = 0;
```

Comment:
Style nit: ` = 0` is redundant. Also these could be separately declared where they are used, but that's a matter of preference.